core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/heuristic/selector/move/generic/list/kopt/TimSort.java

```diff
@@ -0,0 +1,900 @@
+package org.optaplanner.core.impl.heuristic.selector.move.generic.list.kopt;
+
+import java.util.Comparator;
+
+/**
+ * Adapted from OpenJDK's {@link java.util.Arrays#parallelSort(Object[], int, int, Comparator)}
+ * to support primitive int arrays.
+ *
+ * TODO smells a bit of plagiarism; maybe we should bring our own sort?
+ * I'm thinking block sort (https://en.wikipedia.org/wiki/Block_sort)
+ * Stable, O(n log n) worst case, no extra memory use.
+ * That said, TimSort already does all of that except for the memory use.
```

Comment:
I think I have a solution to the problem:
https://fastutil.di.unimi.it/docs/it/unimi/dsi/fastutil/ints/IntArrays.html#stableSort(int%5B%5D,int,int,it.unimi.dsi.fastutil.ints.IntComparator)

Let's include the `fastutil-core` library. I've been meaning to do it anyway, as there are some places in our core where the fast collections will be beneficial. This will at least force me to finally look into it. :-)